test/lit/passes/duplicate-function-elimination_all-features.wast

```diff
@@ -0,0 +1,80 @@
+;; NOTE: Assertions have been generated by update_lit_checks.py --all-items and should not be edited.
+;; NOTE: This test was ported using port_test.py and could be cleaned up.
+
+;; RUN: foreach %s %t wasm-opt --duplicate-function-elimination --all-features -S -o - | filecheck %s
+
+;; --duplicate-function-elimination should not remove functions used in ref.func.
+(module
+  ;; CHECK:      (type $none_=>_i32 (func (result i32)))
+
+  ;; CHECK:      (type $none_=>_funcref (func (result funcref)))
+
+  ;; CHECK:      (elem declare func $0)
+
+  ;; CHECK:      (func $0 (result i32)
+  ;; CHECK-NEXT:  (i32.const 0)
+  ;; CHECK-NEXT: )
+  (func $0 (result i32)
+    (i32.const 0)
+  )
+  (func $1 (result i32)
+    (i32.const 0)
+  )
+  ;; CHECK:      (func $test (result funcref)
+  ;; CHECK-NEXT:  (ref.func $0)
+  ;; CHECK-NEXT: )
+  (func $test (result funcref)
+    (ref.func $1)
+  )
+)
+;; renaming after deduplication must only affect functions
+(module
+ ;; CHECK:      (type $none_=>_none (func))
+
+ ;; CHECK:      (global $bar i32 (i32.const 0))
+
+ ;; CHECK:      (memory $foo 16 16)
+ (memory $foo 16 16)
+ (global $bar i32 (i32.const 0))
+ ;; CHECK:      (export "memory" (memory $foo))
+ (export "memory" (memory $foo))
+ ;; CHECK:      (export "global" (global $bar))
+ (export "global" (global $bar))
+ ;; CHECK:      (func $bar
+ ;; CHECK-NEXT:  (nop)
+ ;; CHECK-NEXT: )
+ (func $bar ;; happens to share a name with the global
+ )
+ (func $foo ;; happens to share a name with the memory
+ )
+)
+;; renaming after deduplication must update ref.funcs in globals
+(module
+ ;; CHECK:      (type $func (func (result i32)))
+ (type $func (func (result i32)))
+ ;; CHECK:      (global $global$0 (ref $func) (ref.func $foo))
+ (global $global$0 (ref $func) (ref.func $bar))
+ ;; These two identical functions can be merged. The ref.func in the global must
+ ;; be updated accordingly.
+ ;; CHECK:      (export "export" (func $2))
+
+ ;; CHECK:      (func $foo (result i32)
+ ;; CHECK-NEXT:  (unreachable)
+ ;; CHECK-NEXT: )
+ (func $foo (result i32)
+  (unreachable)
+ )
+ (func $bar (result i32)
+  (unreachable)
+ )
```

Comment:
The function `$bar` is removed by this optimization, but the CHECKs don't make that obvious. Even just the fact that it has no CHECKs isn't quite enough, as the function after it, "export", also has no CHECKs - its CHECKs actually appear after the module.

Could we emit a CHECK that says a function no longer exist?

Separately, adding a `$` name for the final function, and not just an export name, would help things like this.